modules/instance_template/main.tf

```diff
@@ -43,13 +43,18 @@ locals {
   # initialize the block only if it is enabled.
   shielded_vm_configs = var.enable_shielded_vm ? [true] : []
 
-  gpu_enabled            = var.gpu != null
-  alias_ip_range_enabled = var.alias_ip_range != null
+  gpu_enabled                    = var.gpu != null
+  alias_ip_range_enabled         = var.alias_ip_range != null
+  snp_confidential_instance_type = var.confidential_instance_type == "SEV_SNP"
   on_host_maintenance = (
-    var.preemptible || var.enable_confidential_vm || local.gpu_enabled || var.spot
+    var.preemptible || var.enable_confidential_vm || local.gpu_enabled || var.spot || local.snp_confidential_instance_type
     ? "TERMINATE"
     : var.on_host_maintenance
   )
+
+  # must be set to "AMD Milan" if confidential_instance_type is set to "SEV_SNP", or this will fail to create the VM.
+  min_cpu_platform = local.snp_confidential_instance_type ? "AMD Milan" : var.min_cpu_platform
```

Comment:
I'm not sure if hard-coding "AMD Milan" to `min_cpu_platform` is the best approach. This could lead to non-expected behaviors when user doesn't notice/know this rule and passes and expected another value to `var.min_cpu_platform`. 

How about implement a [variable validation](snp_confidential_instance_type) instead? ([See this validation example being used in foundation](https://github.com/terraform-google-modules/terraform-example-foundation/blob/0a3e97714e6f5c3a6ebfb4704b69b0bffe7773a9/0-bootstrap/modules/parent-iam-member/variables.tf#L21))

Reply:
Actually, I just noticed that there are validation examples in `insteance_template` module. [See here.](https://github.com/terraform-google-modules/terraform-google-vm/blob/7ed37800056e7c53313803fb1d926c37eee232bf/modules/instance_template/variables.tf#L110)

Reply:
As talked offline, we decided to keep the way it is.